cmd/ipam-node/main.go

```diff
@@ -13,7 +13,313 @@
+		return nil, fmt.Errorf("cannot create kubeconfig temp file: %v", err)
+	}
+
+	templateKubeconfig, err := template.New("kubeconfig").Parse(kubeConfigTemplate)
+	if err != nil {
+		return nil, fmt.Errorf("template parse error: %v", err)
+	}
+	templateData := map[string]string{
+		"KubeConfigHost":          fmt.Sprintf("%s://[%s]:%s", kubeProtocol, kubeHost, kubePort),
+		"KubeServerTLS":           tlsConfig,
+		"KubeServiceAccountToken": string(saTokenByte),
+	}
+
+	// Prepare
+	hash := sha256.New()
+	writer := io.MultiWriter(hash, fp)
+
+	// genearate kubeconfig from template
+	if err = templateKubeconfig.Execute(writer, templateData); err != nil {
+		return nil, fmt.Errorf("cannot create kubeconfig: %v", err)
+	}
+
+	if err := fp.Sync(); err != nil {
+		os.Remove(fp.Name())
+		return nil, fmt.Errorf("cannot flush kubeconfig temp file: %v", err)
+	}
+	if err := fp.Close(); err != nil {
+		os.Remove(fp.Name())
+		return nil, fmt.Errorf("cannot close kubeconfig temp file: %v", err)
+	}
+
+	newFileHash := hash.Sum(nil)
+	if currentFileHash != nil && bytes.Equal(newFileHash, currentFileHash) {
+		log.Printf("kubeconfig is same, not copy\n")
+		os.Remove(fp.Name())
+		return currentFileHash, nil
+	}
+
+	// replace file with tempfile
+	if err := os.Rename(tempKubeConfigFile, nvIPAMKubeConfig); err != nil {
+		return nil, fmt.Errorf("cannot replace %q with temp file %q: %v", nvIPAMKubeConfig, tempKubeConfigFile, err)
+	}
+
+	log.Printf("kubeconfig is created in %s\n", nvIPAMKubeConfig)
+	return newFileHash, nil
+}
+func (o *Options) createNvIpamConfig(currentFileHash []byte) ([]byte, error) {
+	// create kubeconfig by template and replace it by atomic
+	tempNvIpamConfigFile := fmt.Sprintf("%s/nv-ipam.d/nv-ipam.conf.new", o.CNIConfDir)
+	nvIpamConfigFile := fmt.Sprintf("%s/nv-ipam.d/nv-ipam.conf", o.CNIConfDir)
+	fp, err := os.OpenFile(tempNvIpamConfigFile, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0600)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create nv-ipam.conf temp file: %v", err)
+	}
+
+	templateNvIpamConfig, err := template.New("nv-ipam-config").Parse(nvIpamConfigTemplate)
+	if err != nil {
+		return nil, fmt.Errorf("template parse error: %v", err)
+	}
+
+	templateData := map[string]string{
+		"KubeConfigFile": o.NvIpamKubeConfigFileHost,
+		"NvIpamDataDir":  o.NvIpamCNIDataDirHost,
+		"NvIpamLogFile":  o.NvIpamLogFile,
+		"NvIpamLogLevel": o.NvIpamLogLevel,
+	}
+
+	// Prepare
+	hash := sha256.New()
+	writer := io.MultiWriter(hash, fp)
+
+	// genearate nv-ipam-config from template
+	if err = templateNvIpamConfig.Execute(writer, templateData); err != nil {
+		return nil, fmt.Errorf("cannot create nv-ipam-config: %v", err)
+	}
+
+	if err := fp.Sync(); err != nil {
+		os.Remove(fp.Name())
+		return nil, fmt.Errorf("cannot flush nv-ipam-config temp file: %v", err)
+	}
+	if err := fp.Close(); err != nil {
+		os.Remove(fp.Name())
+		return nil, fmt.Errorf("cannot close nv-ipam-config temp file: %v", err)
+	}
+
+	newFileHash := hash.Sum(nil)
```

Comment:
hash check mechanism is dead code currently.
also orig hash can be part of same method and not passed as arg from caller.

anyway, not relevant for now.